packages/hierarchy-builder/src/hierarchy-builder/internal/operators/grouping/PropertiesGrouping.ts

```diff
@@ -0,0 +1,323 @@
+}
+
+/** @internal */
+export interface PropertyGroupInfo {
+  ecClass: ECClass;
+  previousPropertiesGroupingInfo: PreviousPropertiesGroupingInfo;
+  propertyGroup: Omit<PropertyGroup, "propertyValue">;
+}
+
+/** @internal */
+export type PreviousPropertiesGroupingInfo = Array<{ fullClassName: string; propertyGroup: Omit<PropertyGroup, "propertyValue"> }>;
+
+/** @internal */
+export async function createPropertyGroups(
+  metadata: IMetadataProvider,
+  nodes: ProcessedInstanceHierarchyNode[],
+  extractedPropertyInfo: PropertyGroupInfo,
+  valueFormatter: IPrimitiveValueFormatter,
+): Promise<GroupingHandlerResult> {
+  const groupings: PropertyGroupingInformation = { ungrouped: [], grouped: new Map() };
+  for (const node of nodes) {
+    const byProperties = node.processingParams?.grouping?.byProperties;
+    if (!byProperties) {
+      groupings.ungrouped.push(node);
+      continue;
+    }
+    if (!(await shouldCreatePropertyGroup(metadata, extractedPropertyInfo, byProperties, node.key.instanceKeys[0].className))) {
+      groupings.ungrouped.push(node);
+      continue;
+    }
+    const currentProperty = byProperties.propertyGroups[extractedPropertyInfo.previousPropertiesGroupingInfo.length];
+    const partialPropertyNodeKeyToAdd = {
+      propertyName: currentProperty.propertyName,
+      fullClassName: byProperties.fullClassName,
+    };
+
+    if (currentProperty.propertyValue === undefined || currentProperty.propertyValue === "") {
+      if (byProperties.createGroupForUnspecifiedValues) {
+        addGroupingToMap(
+          groupings.grouped,
+          `${currentProperty.propertyName}:Unspecified`,
+          {
+            label: translate("grouping.unspecified-label"),
+            propertyGroupingNodeKey: {
+              ...partialPropertyNodeKeyToAdd,
+              type: "property-grouping:value",
+              formattedPropertyValue: "",
+            },
+          },
+          node,
+        );
+        continue;
+      }
+      groupings.ungrouped.push(node);
+      continue;
+    }
+
+    if (currentProperty.ranges) {
+      if (typeof currentProperty.propertyValue === "number") {
+        const propValue = currentProperty.propertyValue;
+        const matchingRange = currentProperty.ranges.find((range) => propValue >= range.fromValue && propValue <= range.toValue);
+
+        if (matchingRange) {
+          const fromValueTypedPrimitive = {
+            type: Number.isInteger(matchingRange.fromValue) ? "Integer" : "Double",
+            value: matchingRange.fromValue,
+          } as TypedPrimitiveValue;
+          const toValueTypedPrimitive = {
+            type: Number.isInteger(matchingRange.toValue) ? "Integer" : "Double",
+            value: matchingRange.toValue,
+          } as TypedPrimitiveValue;
+
+          const rangeLabel = matchingRange.rangeLabel ?? `${await valueFormatter(fromValueTypedPrimitive)} - ${await valueFormatter(toValueTypedPrimitive)}`;
```

Comment:
The property may have extra information that affects value formatting: extended type name, kind of quantity. I think "from" and "to" should be formatted with that extra information included, similar to how we format property value (the non-ranges case)

Reply:
Changed to your suggestion, now it is done like this:
```ts
const fromValueTypedPrimitive = {
     type: Number.isInteger(matchingRange.fromValue) ? "Integer" : "Double",
     extendedType: property.extendedTypeName,
     koqName: (await property.kindOfQuantity)?.fullName,
     value: matchingRange.fromValue,
} as TypedPrimitiveValue;
const toValueTypedPrimitive = {
     type: Number.isInteger(matchingRange.toValue) ? "Integer" : "Double",
     extendedType: property.extendedTypeName,
     koqName: (await property.kindOfQuantity)?.fullName,
     value: matchingRange.toValue,
} as TypedPrimitiveValue;

const rangeLabel = matchingRange.rangeLabel ?? `${await valueFormatter(fromValueTypedPrimitive)} - ${await valueFormatter(toValueTypedPrimitive)}`;
```